redbot/core/utils/__init__.py

```diff
@@ -424,3 +426,67 @@ async def without_duplicates(self) -> AsyncIterator[_T]:
                 yield item
                 _temp.add(item)
         del _temp
+
+    async def find(
+        self, predicate: Union[Callable, Coroutine], default: Optional[Any] = None
+    ) -> AsyncIterator[_T]:
+        """Calls ``predicate`` over items in iterable and return first value to match.
+
+        Parameters
+        ----------
+        predicate: Union[Callable, Coroutine]
+            A function that returns a boolean-like result. The predicate provided can be a coroutine.
+        default: Optional[Any]
+            The value to return if there is no matches.
+
+        Raises
+        ------
+        TypeError
+            When ``predicate`` is not a callable.
+
+        Examples
+        --------
+        >>> from redbot.core.utils import AsyncIter
+        >>> await AsyncIter(range(3)).find(lambda x: x == 1)
+        1
+        """
+        while True:
+            try:
+                elem = await self.__anext__()
+            except StopAsyncIteration:
+                return default
+            ret = await maybe_coroutine(predicate, elem)
+            if ret:
+                return elem
+
+    def map(self, func: Union[Coroutine[..., _S], Callable[..., _S]]) -> AsyncIter[_S]:
```

Comment:
Type hint should specify what should the function `func` accept:
```suggestion
    def map(self, func: Callable[[_T], Union[_S, Awaitable[_S]]]) -> AsyncIter[_S]:
```